core/embed/rust/src/ui/geometry.rs

```diff
@@ -329,11 +344,26 @@ impl Rect {
         self.bottom_right().center(self.top_right())
     }
 
+    pub fn is_empty(&self) -> bool {
+        self.x0 >= self.x1 || self.y0 >= self.y1
+    }
+
     /// Whether a `Point` is inside the `Rect`.
     pub const fn contains(&self, point: Point) -> bool {
         point.x >= self.x0 && point.x < self.x1 && point.y >= self.y0 && point.y < self.y1
     }
 
+    pub fn has_intersection(&self, r: Rect) -> bool {
+        self.x0 < r.x1 && self.x1 > r.x0 && self.y0 < r.y1 && self.y1 > r.y0
+    }
+
+    pub fn intersect(&self, r: Rect) -> Rect {
+        Rect::new(
```

Comment:
this seems to return a 'negative' rectangle in case of no intersection. I would more expect 'zero' rectangle of even returning `Option<Rect>`

Reply:
A "negative" rectangle is actually an empty rectangle. When applying intersection or union operations on negative rectangles, they behave as if they were a "zero" rectangle. Therefore, this does not lead to undefined behavior. While we may return a "zero" rectangle, it's not without a cost; thus, I would prefer not to do so. Returning an `Option<Rect>` is highly impractical.

Reply:
While it is true for unions and intersect operations on rectangle, its does not behave the same for other operations, for example `height()` will return different result for zero or negative rectangle. From there, we would need to be sure that every algorithm using also height is correctly handling negative results, otherwise bugs might be introduced.

Regarding the Option, i would say it can be also more practical in _some_ cases, i.e. 
```
if let Some(intersect) = x.intersect(y) {
 do something with intersect
}
```
is more convenient then
```
if x.has_intersection(y) {
 do something with x.intersect(y)
}
```

But it perhaps depends mostly on the intended usage which i don't see here, so i will leave that up to you.

If this is performance critical you can leave it as is, but consider adding comment about this.

Reply:
I actually understand your point of view, but the issue is more about defining what determines an empty rectangle and how we interpret negative values for height or width. It's quite easy to create a "negative" rectangle through  new(), inset(), or with_xxx() functions. and I think it's essential for us to acknowledge that negative width or height isn't inherently abnormal and should be handled accordingly. 

So I decided to just add few comments to make it more clear, see d6a575e88629cbae10fc5df9dc6e66b6c13d435f

But I am open to further discussion:-)

Reply:
I agree with @TychoVrahe. Negative rectangles seem to be a problem -- in particular, `normalize()` will magically turn a negative rectangle into a normal one, which seems like a footgun.

...what is the use of `normalize()` anyway? it looks like a magnet for surprises

if we need super high performance for `intersect()`, at least call it `intersect_unchecked` and add an `intersect` method that does "if !self.has_intersection(other) { return Self::empty() } else self.intersect_unchecked(other)"

Reply:
I agree that `normalize()` looks hacky. Its purpose is a bit special, and it's only used in one place. It helps create rectangles from two arbitrary points, regardless of which corner of the rectangle they are. Perhaps it would be clearer to have a function `Rect::from_arbitrary_points(Point, Point)` or to remove it altogether.

Reply:
After finding the `clamp()` function and discussing it in person, I've made the following changes:
1) removed `normalize()` as it will be replaced by something else where it's currently used
2) rermoved `intersect()` - we will continue using `clamp()` instead.
3) removed `has_intersection()` - we will use `(!intersect().is_empty)` instead.

see a04980460dd50c960648c19c6ee5c30e55cb9fd6